scheduler/src/cook/scheduler/scheduler.clj

```diff
@@ -686,46 +686,25 @@
                              (map tools/job->usage)
                              (reduce (partial merge-with +))))))))
 
-; This is used by the /unscheduled_jobs code to determine whether
-; or not to report rate-limiting as a reason for being pending
-(defonce pool->user->num-rate-limited-jobs (atom {}))
-
 (defn pending-jobs->considerable-jobs
   "Limit the pending jobs to considerable jobs based on usage and quota.
-   Further limit the considerable jobs to a maximum of num-considerable jobs."
+   Further limit the considerable jobs to a maximum of num-consideracheckble jobs."
   [db pending-jobs user->quota user->usage num-considerable pool-name]
   (log/debug "In" pool-name "pool, there are" (count pending-jobs) "pending jobs:" pending-jobs)
-  (let [enforcing-job-launch-rate-limit? (ratelimit/enforce? ratelimit/job-launch-rate-limiter)
-        user->number-jobs (atom {})
-        user->rate-limit-count (atom {})
-        user-within-launch-rate-limit?-fn
-        (fn
-          [{:keys [job/user]}]
-          ; Account for each time we see a job for a user.
-          (swap! user->number-jobs update user #(inc (or % 0)))
-          (let [tokens (ratelimit/get-token-count! ratelimit/job-launch-rate-limiter user)
-                number-jobs-for-user-so-far (@user->number-jobs user)
-                is-rate-limited? (> number-jobs-for-user-so-far tokens)]
-            (when is-rate-limited?
-              (swap! user->rate-limit-count update user #(inc (or % 0))))
-            (not (and is-rate-limited? enforcing-job-launch-rate-limit?))))
+  (let [enforcing-job-launch-rate-limit? (ratelimit/enforce? quota/per-user-per-pool-launch-rate-limiter)
         considerable-jobs
         (->> pending-jobs
              (tools/filter-pending-jobs-for-quota pool-name user->quota user->usage (tools/global-pool-quota (config/pool-quotas) pool-name))
              (filter (fn [job] (tools/job-allowed-to-start? db job)))
-             (filter user-within-launch-rate-limit?-fn)
              (filter launch-plugin/filter-job-launches)
              (take num-considerable)
              ; Force this to be taken eagerly so that the log line is accurate.
-             (doall))]
-    (swap! pool->user->num-rate-limited-jobs update pool-name (constantly @user->rate-limit-count))
+             (doall))
+        user->rate-limit-count (get @tools/pool->user->num-rate-limited-jobs pool-name (atom {}))]
```

Comment:
Why do the values in this map need to be `atom`s? Isn't it enough that the whole (overall) map is in an `atom`?

Reply:
Good question. There's two levels of atoms here. 1. when we run pending->considerable-jobs, we replace the whole user->rate-limit-count map with an empty one from scratch for that pool. 

The inner atom is, as we go through the lazy list forcing via the doall one line up, we update the inner map depending on which elements in the lazy list got forced. Splitting this way prevents concurrent writing; we can fenzo match different queues in parallel.

Reply:
As discussed, code's been restructured, so this isn't applicable.